incubator/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSessionFactory.java

```diff
@@ -301,29 +433,155 @@ private void onMqttData(
+                        kafkaPayload.sizeof(), flags, kafkaPayload, kafkaDataEx);
+
+                    String16FW willSignalKey = new String16FW(clientId.asString() + WILL_SIGNAL_KEY_POSTFIX);
+                    Flyweight willSignalKafkaDataEx = kafkaDataExRW
+                        .wrap(extBuffer, 0, extBuffer.capacity())
+                        .typeId(kafkaTypeId)
+                        .merged(m -> m
+                            .deferred(0)
+                            .timestamp(now().toEpochMilli())
+                            .partition(p -> p.partitionId(-1).partitionOffset(-1))
+                            .key(b -> b.length(willSignalKey.length())
+                                .value(willSignalKey.value(), 0, willSignalKey.length()))
+                            .hashKey(b -> b.length(clientId.length())
+                                .value(clientId.value(), 0, clientId.length()))
+                            .headersItem(h ->
+                                h.nameLen(TYPE_HEADER_NAME.length())
+                                    .name(TYPE_HEADER_NAME.value(), 0, TYPE_HEADER_NAME.length())
+                                    .valueLen(WILL_SIGNAL_NAME.length())
+                                    .value(WILL_SIGNAL_NAME.value(), 0, WILL_SIGNAL_NAME.length())))
+                        .build();
+
+                    final MqttWillSignalFW willSignal =
+                        mqttWillSignalRW.wrap(willSignalBuffer, 0, willSignalBuffer.capacity())
+                            .clientId(clientId)
+                            .delay(delay)
+                            .deliverAt(MqttWillDeliverAt.UNKNOWN.value())
+                            .lifetimeId(lifetimeId)
+                            .willId(willId)
+                            .instanceId(instanceId.getInstanceId())
+                            .build();
+
+                    willSignalSize = willSignal.sizeof();
+                    session.doKafkaData(traceId, authorization, budgetId, willSignalSize, flags,
+                        willSignal, willSignalKafkaDataEx);
+
+                    evolveSessionStream(traceId, authorization, budgetId);
+                    break;
+                case STATE:
+                    kafkaDataEx = kafkaDataExRW
+                        .wrap(extBuffer, 0, extBuffer.capacity())
+                        .typeId(kafkaTypeId)
+                        .merged(m -> m
+                            .deferred(0)
+                            .timestamp(now().toEpochMilli())
+                            .partition(p -> p.partitionId(-1).partitionOffset(-1))
+                            .key(b -> b.length(clientId.length())
+                                .value(clientId.value(), 0, clientId.length())))
+                        .build();
+
+                    kafkaPayload = mqttSessionStateRO.tryWrap(buffer, offset, limit);
+                    if (kafkaPayload == null)
```

Comment:
We're sending an empty payload from MQTT binding when we have clean start, so no, it's a valid payload right now.

Reply:
Ohh I see. I changed the code a little bit. One thing though: how could we get a malformed sessionState? Can't we assume that since we're the one that's creating the sessionState in the mqtt-binding, it's never malformed?